server/src/api/bill_exec.py

```diff
@@ -52,33 +54,39 @@ def create_bill(self):
         success = False
         
         try:
-            username = cherrypy.session["username"]
+            #Check login
+            try:
+                username = cherrypy.session["username"]
+            except KeyError as e:
+                raise InvalidDataException("login")
+
             #Bill has a title
             if ((len(title) < 1)):
-                raise ValueError("no title","title")
+                raise InvalidDataException("title")
             #Bill has a payer
             if ((len(payer) < 1)):
-                raise ValueError("no payer","payer")
+                raise InvalidDataException("payer")
             #Bill has a total greater than 0
-            if (float(total)<=0):
-                raise ValueError("bad total","total")
+            if (double(total) <= 0): #len(total)==0 or 
+                raise InvalidDataException("total")
             #Bill has someone who needs to pay their share
             if (len(payment_objects) < 1):
-                raise ValueError("bill only has one person","split")
+                raise InvalidDataException("split")
             
             success = self.database.add_bill(payer,username,title,total,payment_objects)
-        except KeyError as e:
-            error_msg = "User not logged in"
-        except Exception as e:
+        except InvalidDataException as e:
             #add the appropriate message to return
-            if(e.args[1]=="title"):
+            print(e)
+            if(e.args[0]=="title"):
                 error_msg = "Title required "
-            elif(e.args[1]=="payer"):
+            elif(e.args[0]=="payer"):
                 error_msg = "No payer for bill "
-            elif(e.args[1]=="total"):
+            elif(e.args[0]=="total"):
                 error_msg = "Incorrect total, total must be greater than 0 "
-            elif(e.args[1]=="split"):
+            elif(e.args[0]=="split"):
                 error_msg = "Cannot split a bill with one person "
+            elif(e.args[0]=="login"):
+                error_msg = "User not logged in"
             else:
                 print(e)
                 error_msg = "Unknown error occured"
```

Comment:
Would it be safer if this part was in a catch-all `except Exception` block instead? At the moment, unintentional exceptions would be silently ignored because of the `finally` clause. E.g., On the app, try splitting a bill with an empty total amount, and the server will return a failure with an empty error message, yet nothing is logged into the console because a type conversion error was silently ignored.

Reply:
It should handle an empty total. I'll add a catch-all which prints to console